plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeLocalConcurrentWritesTest.java

```diff
@@ -167,6 +169,184 @@ private void testConcurrentInsertsReconciliationForBlindInserts(boolean partitio
+            assertTrue(executor.awaitTermination(10, SECONDS));
+        }
+    }
+
+
+    // Copied from BaseDeltaLakeConnectorSmokeTest
+    @Test
+    public void testConcurrentInsertsReconciliationForMixedInserts()
+            throws Exception
+    {
+        int threads = 3;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        ExecutorService executor = newFixedThreadPool(threads);
+        String tableName = "test_concurrent_mixed_inserts_table_" + randomNameSuffix();
+
+        assertUpdate("CREATE TABLE " + tableName + " (a, part) WITH (partitioned_by = ARRAY['part']) AS VALUES (0, 10)", 1);
+
+        try {
+            // insert data concurrently
+            executor.invokeAll(ImmutableList.<Callable<Void>>builder()
+                            .add(() -> {
+                                barrier.await(10, SECONDS);
+                                // Independent of the execution order, use on the non-blind insert a partition constraint which is not `TupleDomain.all()`
+                                // which should lead to reconciliation
+                                getQueryRunner().execute("INSERT INTO " + tableName + " SELECT COUNT(*) AS a, part FROM " + tableName + " WHERE part = 10 GROUP BY part");
+                                return null;
+                            })
+                            .add(() -> {
+                                barrier.await(10, SECONDS);
+                                getQueryRunner().execute("INSERT INTO " + tableName + " VALUES (11, 20)");
+                                return null;
+                            })
+                            .add(() -> {
+                                barrier.await(10, SECONDS);
+                                getQueryRunner().execute("INSERT INTO " + tableName + " VALUES (22, 30)");
+                                return null;
+                            })
+                            .build())
+                    .forEach(MoreFutures::getDone);
+
+            assertQuery(
+                    "SELECT * FROM " + tableName,
+                    "VALUES (0, 10), (1, 10), (11, 20), (22, 30)");
+            assertQuery(
+                    "SELECT version, operation, isolation_level, read_version FROM \"" + tableName + "$history\"",
+                    """
+                            VALUES
+                                (0, 'CREATE TABLE AS SELECT', 'WriteSerializable', 0),
+                                (1, 'WRITE', 'WriteSerializable', 0),
+                                (2, 'WRITE', 'WriteSerializable', 1),
+                                (3, 'WRITE', 'WriteSerializable', 2)
+                            """);
+        }
+        finally {
+            assertUpdate("DROP TABLE " + tableName);
+            executor.shutdownNow();
+            assertTrue(executor.awaitTermination(10, SECONDS));
+        }
+    }
+
+    @Test
+    public void testConcurrentInsertsSelectingFromDifferentPartitionsOfSameTable()
```

Comment:
Any reason why this test is not in `BaseDeltaLakeConnectorSmokeTest` as well?

Reply:
I'm reluctant that there is added value to test extensively on  `BaseDeltaLakeConnectorSmokeTest`.
The class would not be a "smoke" test class anymore.

I trust that the `local` file store does indeed throw `FileAlreadyExistsException` when a concurrent operation creates a namesake transaction log file and that's why I'd be inclined to say `TestDeltaLakeLocalConcurrentWritesTest` is fit to hold an increased battery of tests for concurrent writes reconciliation functionality.